src/osd/PrimaryLogPG.cc

```diff
@@ -2385,17 +2385,28 @@ int PrimaryLogPG::do_manifest_flush(OpRequestRef op, ObjectContextRef obc, Flush
       bufferlist in;
       if (fp_oid != tgt_soid.oid) {
 	// decrement old chunk's reference count 
-	ObjectOperation dec_op;
-	cls_chunk_refcount_put_op put_call;
-	put_call.source = soid;
-	::encode(put_call, in);                             
-	dec_op.call("cas", "chunk_put", in);         
-	// we don't care dec_op's completion. scrub for dedup will fix this.
-	tid = osd->objecter->mutate(
-	  tgt_soid.oid, oloc, dec_op, snapc,
-	  ceph::real_clock::from_ceph_timespec(obc->obs.oi.mtime),
-	  flags, NULL);
-	in.clear();
+	SnapSetContext* ssc = obc->ssc;
+	bool has_reference = false;
+	if (ssc && ssc->snapset.clones.size() > 0) {
+	  if (check_has_ref_snap(soid, ssc->snapset.clone_overlap.rbegin()->first,
```

Comment:
Hmm, I don't think it's necessarily the case that we're flushing the head here.  Can't this be called on a snapshot?

Reply:
I don't entirely understand the logic here.  Can an object be flushed if it currently references chunks?  If it currently is chunked, wouldn't that necessarily mean it's clean?

Reply:
@athanatos The purpose of this PR is to prevent removing referenced chunks. (https://lists.ceph.io/hyperkitty/list/dev@ceph.io/thread/7G222W5F7MD7X5MJS3BF6YDJ76RKJGN5/)
Flush happens when some chunks in an object are dirty. 
At this time, if object's chunk A has different contents from before, chunks A's reference count is decreased and then, a new chunk needs to be stored. 
In the logic here, it checks whether previous snapshots refer to the chunks. If it has a reference, skip the chunk_put (reference decrement).
This is because reference counted chunks.  For instance, the object (which has chunks) is deleted, which results in dropping the reference and then, the chunk will be removed 
if chunk's reference count is zero.

an object has chunks.
a chunked object has the reference.

Reply:
Ah, so we don't release the ref when we dirty the object, we wait until we flush the dirty chunks.

Reply:
Why is it checking for overlap with the last snapshot?  If we're flushing not the head, wouldn't it need to check the next smallest one?

Reply:
Because overlap means that there is references so, we shouldn't flush the object. If we're flushing not the head, this can be a solution. But, it prevents flushing the object that can be flushed.

Reply:
Suppose the current snapset for an object is 13:[12(11, 9), 6(5)]+head

If clones 12 and 6 are clean and we're flushing head, then yes, clone_overlap.rbegin() (which is the overlap between cloneid 12 and head) is the correct overlap to check.

However, suppose clone 12 is still dirty.  We need to flush clone id 12.  In that case, we also must consider the overlap between 12 and 6, which would be --clone_overlap.rbegin().  That is, this code seems to only be correct for head, but it appears to me that it can flush clones as well.  Am I reading this right?  Do you have a test case for the situation I'm describing?

Reply:
Right. I couldn't consider your concern. This PR is based on following concepts.

1) If a newly dereferenced (by head) chunk overlaps with the most recent
clone, do nothing--that clone still has a reference to it.

2) If a newly dereferenced (by head) chunk does NOT overlap with the most
recent clone, then it is the only referant, and we can decrement it after
we apply the update (like we do today).

I will add a test case and patches.

Reply:
So you are adding code to handle the case where it's a clone and you need to consider the overlap with the next and previous clones?

Reply:
Not yet. I'm looking for how to implement a test case you described and how to add code to handle the case. After gathering all your comments, I'll start coding. Any other comments?